10 - More Declarative JSX/01 - Complex Structure.js

```diff
@@ -0,0 +1,94 @@
+
+
+// This feature does NOT propose the addition of special templating tags.  It only introduces the ability for tags to bind
+// variables, thereby allowing custom components that can be composed in a way that is more readable and feels more declarative.
+// We assume implementations of "push", "if" and "foreach" tags.  The actual implementation is quite trivial, and
+// implementable by the end user.
+
+
+// Currently, to render a bookstore, you would need the following code...
+var booksData = [{title: "Harry Potter", authors: ["JK Rowling"], ...}, ...];
+return <div>{booksData.map(function(bookData){
+  var authorElements = [];
+  bookData.authors.foreach(function(authorData, idx){
+    authorElements.push(<a href={"http://www.authors.com?id="+author.id} style={{color: 'blue', textDecoration: 'none'}}>{author.name}</a>);
+    if(idx !== bookData.authors.length) authorElements.push(", ");
+  });
+  return <div>
+    <h1>{bookData.title}</h1>
+    By: 
+    { bookData.authors.length === 0 ? <span style={{color: 'gray'}}>Unknown author</span> :  {authorElements}}
+  </div>;
+})}</div>;
+
+
+// The problem is that the code above is complex and hard to follow.  it feels a lot like back in the days when
+// user interfaces were built using an imperative API (like IOS, Swing, etc).
+// In order to generate the comma-separated list of authors for each book, we had to create a new array, iterate over
+// the data, and push elements into the new array, and subsequently put the array of elements into the markup.
+// This was a simple example, but you can imagine this widget becoming much more complicated, and it becoming hard
+// to track what's happening in the user interface
+
+// Now let's try something new.
+
+// Ideally, you could do the same thing in a less imperative way by evaluating variables lazily...
+<push name="books" value={[{title: "Harry Potter", authors: ["JK Rowling"], ...}, ...]}>
+  <div>
+    <foreach element="book" collection={books}>
+      <div>
+        <h1>{title}</h1>
+        By:
+        <iff test={book.authors.length === 0}>
+          <span style={{color: 'gray'}}>Unknown author</span>
+        </iff>
+        <iff test={book.authors.length > 0}>
+          <foreach element="author" index="authorIndex" collection={book.authors} />
+            <a
+              href={"http://www.authors.com?id="+author.id}
+              style={{color: 'blue', textDecoration: 'none'}}>{author.name}
```

Comment:
There are a couple of options for compiling JSX to JS, none of them particularly fantastic, but all workable.  One would be to compile `author.id` to `$scope.author.id`.  Another would be to use the `with` operator.

It's worth noting that the expressions will probably need to be wrapped in a function/closure and executed during render instead of during element creation, but that's not a big deal, should be virtually invisible, though there is a slight performance overhead of creating/invoking the anonymous function.  This can probably be minimized by checking which variables are in scope and only converting to a function if you are using variables inherited from a tag.

Regardless of the exact desugaring syntax, it's the concept of allowing tags to bind variables that is most important/interesting.